core/codec/rle/rle_plus_encoding_stream.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CPP_FILECOIN_RLE_PLUS_ENCODING_STREAM_HPP
+#define CPP_FILECOIN_RLE_PLUS_ENCODING_STREAM_HPP
+
+#include <vector>
+#include <set>
+
+#include <boost/dynamic_bitset.hpp>
+
+#include "codec/rle/rle_plus_config.hpp"
+
+namespace fc::codec::rle {
+  /**
+   * @class RLE+ encoding stream
+   */
+  class RLEPlusEncodingStream {
+   public:
+    /**
+     * @brief Encode integer set
+     * @tparam T - type of integer
+     * @tparam A - set's allocator
+     * @param input - content of the set to encode
+     * @return Encoded stream
+     */
+    template <typename T, typename A>
+    RLEPlusEncodingStream &operator<<(const std::set<T, A> &input) {
+      std::vector<T> periods = this->getPeriods(input);
+      this->initContent();
+      bool flag = false;
+      if (!input.empty()) flag = *input.begin() == 0;
+      content_.push_back(flag);
+      for (const auto &value : periods) {
+        if (value == 1) {
+          content_.push_back(true);
+        } else if (value < Config::LONG_BLOCK_VALUE) {
+          this->pushSmallBlock(value);
+        } else if (value >= Config::LONG_BLOCK_VALUE) {
+          this->pushLongBlock(value);
+        }
+      }
+      return *this;
+    }
+
+    /**
+     * @brief Get encoded stream content
+     * @return Stream content
+     */
+    std::vector<uint8_t> data() const {
+      std::vector<uint8_t> data_content;
+      data_content.reserve(content_.num_blocks());
+      boost::to_block_range(content_, std::back_inserter(data_content));
+      return data_content;
+    }
+
+   private:
+    boost::dynamic_bitset<uint8_t> content_{}; /**< RLE+ encoded content */
+
+    /**
+     * @brief Write RLE+ header
+     */
+    void initContent() {
```

Comment:
Move non-template functions to .cpp

Reply:
There are just 6 lines of code in the two non-template functions, I think we don't need additional source code file for this.

Reply:
And they will be copied and compile as many times as many includes we have for this file. I would prefer to create separate cpp, to avoid this